features/steps/utils/geometry.py

```diff
@@ -77,3 +85,41 @@ def is_closed(context, instance):
     precision = get_precision_from_contexts(entity_contexts)
     points_coordinates = get_points(instance)
     return math.dist(points_coordinates[0], points_coordinates[-1]) < precision
+
+
+def evaluate_segment(segment: ifcopenshell.entity_instance, dist_along: float) -> np.ndarray:
+    s = ifcos_geom.settings()
+    pwf = wrapper.map_shape(s, segment.wrapped_data)
+
+    prev_trans_matrix = pwf.evaluate(dist_along)
+
+    return np.array(prev_trans_matrix, dtype=np.float64).T
+
+def alignment_segment_positional_difference(length_unit_scale_factor, previous_segment, segment_to_analyze):
+
+    u = abs(float(previous_segment.SegmentLength.wrappedValue)) * length_unit_scale_factor
+    prev_end_transform = evaluate_segment(segment=previous_segment, dist_along=u)
+
+    pX = prev_end_transform[3][0] / length_unit_scale_factor
+    pY = prev_end_transform[3][1] / length_unit_scale_factor
+    preceding_end = (pX, pY)
+
+    current_start = segment_to_analyze.Placement.Location.Coordinates
```

Comment:
1 How sure are we Placement.Location is 2D. I assume it would mess up our call to dist() if we pass a 2d and a 3d vector.
2 How sure are we Placement is not Linear?
3 Did you consider simply calling evaluate(0)?

Reply:
> 1 How sure are we Placement.Location is 2D. I assume it would mess up our call to dist() if we pass a 2d and a 3d vector. 

Good point, I have made this explicit.

> 2 How sure are we Placement is not Linear?

Not sure I understand.  Do you mean that `Placement` might be of type `IfcAxis2PlacementLinear`?  And I assume that this would be for things like an alignment defined by IfcOffsetCurveByDistances?

> 3 Did you consider simply calling evaluate(0)?

Yes, in fact I initially had the logic set up that way but prefer it as-is as it has the side-effect of providing and double-check on imperial unit conversions.  It was a happy accident that I started developed with an imperial example, otherwise I might have completely overlooked this.

Reply:
>  Do you mean that Placement might be of type IfcAxis2PlacementLinear?

yes, this is allowed by the schema, or am I hallucinating this time?

Reply:
It appears to be allowed by the schema.  I am not aware of any examples in the wild that are using this.

Reply:
To resolve this part of the conversation - I recommend that we leave the rule as-is and see if any implementations actually utilize `IfcAxis2PlacementLinear` before we spend effort handling that potential wrinkle.

In other words, not letting the perfect be the enemy of the good :smiley:

Reply:
> In other words, not letting the perfect be the enemy of the good 😃

But the problem is that when it occurs it's not gracefully/explicitly handled. Even if we don't "support" this configuration for checking the rule, I don't think that we should let the check crash and burn on an AttributeError.

Because even if we later add another rule that says this Placement should be local, than still we cannot raise exceptions here, because it would prevent the other rule results from showing up.

I would say:

- We do an if-statement here just to prevent the exception, but doesn't need to be handled here
- We ask for feedback in the IF whether we can constrain the schema here with another normative rule. These kind of "imprecise" attributes can be real interoperability killers.

Reply:
So in summary you are correct, of course, given your wide understanding of potential land mines in the schema.  The best solution is to use `evaluate(u=0.0)` to determine the start position and direction angle of the segment being analyzed.  Which was one of your first questions in review of this PR.  So you get a double gold star.